src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs

```diff
@@ -1944,45 +1943,89 @@ public StringBuilder Replace(string oldValue, string? newValue, int startIndex,
+                // find all occurrences of it contained entirely within the chunk. We stop searching
+                // once we're within oldValue.Length from the end of the chunk (or count limit), at which point
+                // we need to consider a value that bridges between two chunks.
+                ReadOnlySpan<char> remainingChunk = chunk.m_ChunkChars.AsSpan(indexInChunk, Math.Min(chunk.m_ChunkLength - indexInChunk, count));
+                while (oldValue.Length <= remainingChunk.Length)
                 {
-                    // Push it on the replacements array (with growth), we will do all replacements in a
-                    // given chunk in one operation below (see ReplaceAllInChunk) so we don't have to slide
-                    // many times.
-                    if (replacementsCount >= replacements.Length)
+                    // Find the next match.
+                    int foundPos = remainingChunk.IndexOf(oldValue);
+                    if (foundPos >= 0)
                     {
-                        int[] tmp = new int[replacements.Length * 3 / 2 + 4]; // Grow by ~1.5x, but more in the beginning
-                        replacements.CopyTo(tmp);
-                        replacements = tmp;
+                        // We found one.  Add it as a location for the replacement.
+                        indexInChunk += foundPos;
+                        replacements.Append(indexInChunk);
+
+                        // Move ahead to the next location.
+                        remainingChunk = remainingChunk.Slice(foundPos + oldValue.Length);
+                        indexInChunk += oldValue.Length;
+                        count -= foundPos + oldValue.Length;
+
+                        // If after accounting for moving past the match our count has
+                        // gone to 0, break out to stop searching.
+                        Debug.Assert(count >= 0, "count should never go negative");
+                        if (count == 0)
+                        {
+                            break;
+                        }
+                    }
+                    else
+                    {
+                        // No match found. Reposition to one character beyond the last starting
+                        // location searched, which will be oldValue.Length - 1 from the end.
+                        // Then break out so that we can start the cross-chunk matching from that location.
+                        int move = remainingChunk.Length - (oldValue.Length - 1);
+                        indexInChunk += move;
+                        count -= move;
+                        break;
                     }
-                    replacements[replacementsCount++] = indexInChunk;
-                    indexInChunk += oldValue.Length;
-                    count -= oldValue.Length;
                 }
-                else
+
+                Debug.Assert(oldValue.Length > chunk.m_ChunkLength - indexInChunk || count == 0,
```

Comment:
This assert has failed in https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-81098-merge-25df720c9d4e4a9683/System.Runtime.Tests/1/console.4fb7bee6.log?helixlogtype=result:

```log
  Starting:    System.Runtime.Tests (parallel test collections = on, max threads = 2)
Process terminated. Assertion failed.
oldValue.Length = 3, chunk.m_ChunkLength - indexInChunk = 14, count == 2
   at System.Text.StringBuilder.Replace(String oldValue, String newValue, Int32 startIndex, Int32 count) in /_/src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs:line 1986
   at System.Text.Tests.StringBuilderTests.Replace_String(String value, String oldValue, String newValue, Int32 startIndex, Int32 count, String expected) in /_/src/libraries/System.Runtime/tests/System/Text/StringBuilderTests.cs:line 1727
   at InvokeStub_StringBuilderTests.Replace_String(Object, Object, IntPtr*)
```

Reply:
It's a bogus assert.  Will fix.